pytorch_lightning/trainer/evaluation_loop.py

```diff
@@ -136,6 +140,11 @@ def setup(self, model, max_batches, dataloaders):
         self.max_batches = max_batches
         self.num_dataloaders = self._get_num_dataloaders(dataloaders)
 
+        # wrap user samplers, so we can capture batch indices
+        dataloaders = [LightningBatchSamplerWrapper.to_new_dataloader(dataloader) for dataloader in dataloaders]
```

Comment:
Can we only wrap this if auto_id was requested? because otherwise we won't support custom loaders here, which are not derived from the torch loader which might be a large limitation.

Reply:
`Can we only wrap this if auto_id was requested ?` How would you do this ?

On this one, we already don't support custom DataLoader with DDP if they take custom parameters.

PyTorch Geometric tried to use its NeighborSampler #https://github.com/rusty1s/pytorch_geometric/blob/a2cb48ed24e35615ce53bf88b6c2fc3a7d8d1cc0/torch_geometric/data/sampler.py#L18 with DDP. 

The code broke in there: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/data_loading.py#L122.

I will open a PR to work with him on fix. But it would assume the custom Dataloader need to expose all __init__ parameters as attributes.

But definitely good catch !

Reply:
Need to open a ticket for it.

Reply:
```
UserWarning: Trying to replace your BatchSampler for NeighborSampler dataloader.This would fail as your DataLoader doesn't expose as attributes all its __init__ parameters. Missing attributes are {'edge_index', 'node_idx', 'num_nodes'}
```

Reply:
Works as expected on Pytorch Geometric :)

Reply:
> I will open a PR to work with him on fix. But it would assume the custom Dataloader need to expose all init parameters as attributes.

Yes that's true, but that is a weaker constraint then forcing it to be a certain type.

Reply:
Instead of having `auto_id=True` as an argument in `add_prediction`, we could make it an argument of `test` and then raise a warning/error if the id is not None.